src/main/scala/viper/gobra/ast/frontend/Ast.scala

```diff
@@ -763,17 +763,30 @@ object PGhostifier {
   def unapply[T <: PNode](arg: PGhostifier[T]): Option[T] = Some(arg.actual)
 }
 
-
 /**
   * Specification
   */
+sealed trait PTerminationMeasure extends PNode
```

Comment:
In my opinion, the class structure and the names of classes can be somewhat simplified to avoid repetition and to make the code easier to understand. For example, we could use the following class structure:
```scala
sealed trait PTerminationMeasure extends PNode
sealed trait PInconditionalTerminationMeasure extends PTerminationMeasure
sealed trait PConditionalTerminationMeasureClause extends PNode

case class PConditionalTerminationMeasureIfClause(measure: PInconditionalTerminationMeasure, cond: PExpression) extends PConditionalTerminationMeasureClause

case class PStarMeasure() extends PInconditionalTerminationMeasure with PConditionalTerminationMeasureClause
case class PWildcardMeasure() extends PInconditionalTerminationMeasure
case class PTupleTerminationMeasure(tuple: Vector[PExpression]) extends PInconditionalTerminationMeasure
case class PConditionalTerminationMeasures(clauses: Vector[PConditionalTerminationMeasureClause]) extends PTerminationMeasure
```
where `PStarMeasure` corresponds to `PStarCharacter` in your code and `PWildcardMeasure` corresponds to `PUnderscoreCharacter`. 

Given the time constraints, we can keep the current class structure but we should at least rename the classes as suggested.

Reply:
Alternatively, if we are allowed to parse a list of termination measures instead of an optional (single) termination measure, we can simplify it in the following way
```scala
sealed trait PTerminationMeasure extends PNode
sealed trait PUnconditionalTerminationMeasure extends PTerminationMeasure

case class PStarMeasure() extends PUnconditionalTerminationMeasure 
case class PWildcardMeasure() extends PUnconditionalTerminationMeasure
case class PTupleTerminationMeasure(tuple: Vector[PExpression]) extends PUnconditionalTerminationMeasure
case class PConditionalTerminationMeasure(measure: PUnconditionalTerminationMeasure, cond: PExpression) extends PTerminationMeasure
```

This option requires more well definedness checks, e.g., to rule out cases where we use a conditional termination measure together with an unconditional tuple termination measure. However, it greatly simplifies the code and it is easier to produce better error messages in the well-definedness checker than in the parser.